plugin/rewrite/reverter.go

```diff
@@ -21,22 +22,77 @@ type ResponseRule struct {
 // dig will complain with ';; Question section mismatch: got example.org/HINFO/IN'
 type ResponseReverter struct {
 	dns.ResponseWriter
+	request          *dns.Msg
 	originalQuestion dns.Question
+	originalOPT      *dns.OPT
 	ResponseRewrite  bool
 	ResponseRules    []ResponseRule
 }
 
+// return a deep copy of the OPT record
+func dupOPT(opt *dns.OPT) (o *dns.OPT) {
+	edns0 := make([]dns.EDNS0, len(opt.Option))
+	for i, e := range opt.Option {
+		edns0[i] = e.Copy()
+	}
+	o = &dns.OPT{opt.Hdr, edns0}
+	return o
+}
+
+// return a deep copy of the OPT inside this msg
+func copyOPT(m *dns.Msg) (o *dns.OPT) {
+	for i := len(m.Extra) - 1; i >= 0; i-- {
+		if m.Extra[i].Header().Rrtype == dns.TypeOPT {
+			if opt, ok := m.Extra[i].(*dns.OPT); ok {
+				return dupOPT(opt)
+			}
+		}
+	}
+	return o
+}
+
+// replace the OPT inside the msg with the a copy of the one proposed
+// manage properly if OPT record is nil
+func replaceOPT(m *dns.Msg, o *dns.OPT) {
+	for i := len(m.Extra) - 1; i >= 0; i-- {
+		if m.Extra[i].Header().Rrtype == dns.TypeOPT {
+			if o == nil {
+				copy(m.Extra[i:], m.Extra[i+1:])
+				m.Extra[len(m.Extra)-1] = nil
+				m.Extra = m.Extra[:len(m.Extra)-1]
+				break
+			}
+			m.Extra[i] = dupOPT(o)
+			break
+		}
+	}
+	if o != nil {
+		m.Extra = append(m.Extra, dupOPT(o))
```

Comment:
don't duplicate here, you can duplicate it in outer code

Reply:
yes. changed.